File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    void get_inorder(TreeNode *root,vector<int>&inorder)
    {
        if(!root)
            return;
        
        get_inorder(root->left,inorder);
        inorder.push_back(root->val);
        get_inorder(root->right,inorder);
        
        return;
    }
    
    TreeNode* createTree(vector<int>&inorder,int l,int h)
    {
        if(l>h)
            return NULL;
        
        int mid= l + ((h-l)/2);
        TreeNode *root = new TreeNode(inorder[mid]);
        
        root->left = createTree(inorder,l,mid-1);
        root->right = createTree(inorder,mid+1,h);
        
        return root;       
    }
    
    TreeNode* balanceBST(TreeNode* root) {
        vector<int>inorder;
        get_inorder(root,inorder);
        
        int l=0, h= inorder.size()-1;
        return createTree(inorder,l,h);        
    }
};

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    private ArrayList<Integer>inOrder;
    
    private void getInOrder(TreeNode root){
        
        if(root == null)
            return;
        
        getInOrder(root.left);
        this.inOrder.add(root.val);
        getInOrder(root.right);
        
        return;
    }
    
    private TreeNode createBST(int l, int h){
        
        if(l>h)
            return null;
        
        int mid = l + ((h - l)/2);
        
        TreeNode root = new TreeNode(this.inOrder.get(mid));
        root.left = createBST(l,mid-1);
        root.right = createBST(mid+1,h);
        
        return root;
        
    }
    public TreeNode balanceBST(TreeNode root) {
        
        this.inOrder = new ArrayList<>();
        getInOrder(root);
        int l= 0, h = inOrder.size() -1;
        return this.createBST(0,h);
        
    }
}